### ISSUE TITLE
ICE - rustc panics with "path not fully resolved"

### ISSUE THREAD

**Issue by johnpmayer** [rust-lang/rust#24179]
I tried this code (cargo build)

https://github.com/sebcrozet/nalgebra/tree/8e5ea344365bdd83dd84b0992888e49c22f8b799

The diff from last good is large, but the current goal was a conversion to associated types across the library.

https://github.com/sebcrozet/nalgebra/compare/87ded7fa1e2c51676386b66678beae58183a9d67...8e5ea344365bdd83dd84b0992888e49c22f8b799

```
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'path not fully resolved: PathResolution { base_def: DefSelfTy(2950), last_private: LastMod(AllPublic), depth: 1 }', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/librustc/middle/def.rs:79
```

## Meta

`cargo --version --verbose`:

cargo 0.0.1-pre-nightly (3499c8e 2015-04-05) (built 2015-04-05)

`rustc --version --verbose`:

rustc 1.0.0-nightly (be9bd7c93 2015-04-05) (built 2015-04-06)
binary: rustc
commit-hash: be9bd7c93a53fe554a359e60c5f78991fb783eed
commit-date: 2015-04-05
build-date: 2015-04-06
host: x86_64-unknown-linux-gnu
release: 1.0.0-nightly

Backtrace:

stack backtrace:
   1:     0x7ff4c5274818 - sys::backtrace::write::h4618bbda30a2c1fakHC
   2:     0x7ff4c529fbe7 - panicking::on_panic::h94b32ff0a2406463uUI
   3:     0x7ff4c51d78f3 - rt::unwind::begin_unwind_inner::h310b6bce2d1d1820CzI
   4:     0x7ff4c51d7cdf - rt::unwind::begin_unwind_fmt::h36e3246309d9c41dgyI
   5:     0x7ff4c2e20c86 - middle::def::PathResolution::full_def::h159526808fea34ff7am
   6:     0x7ff4c45ee2f2 - PrivacyVisitor<'a, 'tcx>::check_path::hd2ccc99146e3766514a
   7:     0x7ff4c45ef540 - PrivacyVisitor<'a, 'tcx>.Visitor<'v>::visit_path::h03f0b0222713982edtb
   8:     0x7ff4c45ef3cf - PrivacyVisitor<'a, 'tcx>.Visitor<'v>::visit_item::hab766dca223ddc30Bdb
   9:     0x7ff4c45ef25f - PrivacyVisitor<'a, 'tcx>.Visitor<'v>::visit_item::hab766dca223ddc30Bdb
  10:     0x7ff4c45ef25f - PrivacyVisitor<'a, 'tcx>.Visitor<'v>::visit_item::hab766dca223ddc30Bdb
  11:     0x7ff4c45f651b - check_crate::h3e05585fc295e878FVb
  12:     0x7ff4c58f5859 - driver::phase_3_run_analysis_passes::h360d9d4448a0da27gGa
  13:     0x7ff4c58d92f5 - driver::compile_input::h16213dd19660ff98Qba
  14:     0x7ff4c5990a05 - run_compiler::h2a37c69935ac5f41S4b
  15:     0x7ff4c598e31d - boxed::F.FnBox<A>::call_box::h5013090055299594809
  16:     0x7ff4c598d859 - rt::unwind::try::try_fn::h6394874154857801728
  17:     0x7ff4c531c398 - rust_try_inner
  18:     0x7ff4c531c385 - rust_try
  19:     0x7ff4c598db2b - boxed::F.FnBox<A>::call_box::h8372456613525927705
  20:     0x7ff4c528abd1 - sys::thread::create::thread_start::h0a2c84bef4c472f7MuH
  21:     0x7ff4bf05f0a4 - start_thread
  22:     0x7ff4c4e3acfc - __clone
  23:                0x0 - <unknown>

**Comment by wazboy** [comment]
I have an ICE with a very similar backtrace.

### I tried this code:

```rust
pub trait DecoderState {
    type Error;
}

pub trait Decodable<D: DecoderState> {
    type Error = D::Error;

    fn decode(d: &mut D) -> Result<Self, Self::Error>;
}

struct IAmADecoder;

enum IAmADecoderError {
    BadThings,
}

impl DecoderState for IAmADecoder {
    type Error = IAmADecoderError;
}

struct IAmDecodable;

impl<D: DecoderState> Decodable<D> for IAmDecodable {
    type Error = D::Error;

    fn decode(d: &mut D) -> Result<Self, Self::Error> {
        Ok(IAmDecodable)
    }
}

fn main() {
    let mut d = IAmADecoder;
    IAmDecodable::decode(&mut d);
}
```

### Meta
```
$ rustc --version --verbose
rustc 1.1.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.1.0
```
### Backtrace
```
$ RUST_BACKTRACE=1 rustc compiler_error.rs 
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'path not fully resolved: PathResolution { base_def: DefTyParam(TypeSpace, 0, DefId { krate: 0, node: 19 }, "D"(63)), last_private: LastMod(AllPublic), depth: 1 }', /build/rust/src/rustc-1.1.0/src/librustc/middle/def.rs:81

stack backtrace:
   1:     0x7f0cec9c92d9 - sys::backtrace::write::h534114ba1b06b93294r
   2:     0x7f0cec9d11f9 - panicking::on_panic::h4e97afe0febd2c7biJw
   3:     0x7f0cec991ff2 - rt::unwind::begin_unwind_inner::hfe7dd1cef83ccc0fsow
   4:     0x7f0cec992d87 - rt::unwind::begin_unwind_fmt::h97b246841383a432ynw
   5:     0x7f0cea8364a6 - middle::def::PathResolution::full_def::h6662dd4d69d7f710mjm
   6:     0x7f0ceb4ffb8e - PrivacyVisitor<'a, 'tcx>::check_path::he5062fdd97c5ddddwbb
   7:     0x7f0ceb500c87 - PrivacyVisitor<'a, 'tcx>.Visitor<'v>::visit_path::h63eb4ef2b9ae8266Pzb
   8:     0x7f0ceb500b2f - PrivacyVisitor<'a, 'tcx>.Visitor<'v>::visit_item::hef87e41e5698d047akb
   9:     0x7f0ceb50751b - check_crate::h029a1fe227fad1e5v2b
  10:     0x7f0cecf0f2d8 - driver::phase_3_run_analysis_passes::h7109deb2f0a65b9dtGa
  11:     0x7f0cecef2dec - driver::compile_input::hd8975b759aec0d60Qba
  12:     0x7f0cecfa90e1 - run_compiler::h3b0c3beaef2163aa75b
  13:     0x7f0cecfa6932 - boxed::F.FnBox<A>::call_box::h5202619178778601284
  14:     0x7f0cecfa5ef9 - rt::unwind::try::try_fn::h4053862560305393821
  15:     0x7f0ceca403d8 - rust_try_inner
  16:     0x7f0ceca403c5 - rust_try
  17:     0x7f0cecfa6194 - boxed::F.FnBox<A>::call_box::h4599430653034051152
  18:     0x7f0cec9cff91 - sys::thread::Thread::new::thread_start::h089e987aa4c0e52dzvv
  19:     0x7f0ce7521353 - start_thread
  20:     0x7f0cec621bfc - __clone
  21:                0x0 - <unknown>

```

**Comment by arielb1** [comment]
Fixed in 1.2.0
